server/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java

```diff
@@ -0,0 +1,450 @@
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheEvent;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.utils.ZKPaths;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+
+/**
+ *
+ */
+public class NamespacedExtractionModule implements DruidModule
+{
+  private static final Logger log = new Logger(NamespacedExtractionModule.class);
+  private static final String TYPE_PREFIX = "druid.query.extraction.namespace.cache";
+  private static final String ENABLE_NAMESPACES = "druid.query.extraction.namespace";
+  private final ConcurrentMap<String, Function<String, String>> fnCache = new ConcurrentHashMap<>();
+
+  @Override
+  public List<? extends Module> getJacksonModules()
+  {
+    return ImmutableList.<Module>of(
+        new Module()
+        {
+          @Override
+          public String getModuleName()
+          {
+            return "DruidNamespacedExtractionModule";
+          }
+
+          @Override
+          public Version version()
+          {
+            return Version.unknownVersion();
+          }
+
+          @Override
+          public void setupModule(SetupContext context)
+          {
+            context.registerSubtypes(ExtractionNamespace.class);
+          }
+        }
+    );
+  }
+
+  public static class NamespacedExtractionModuleConfig
+  {
+    @JsonCreator
+    public NamespacedExtractionModuleConfig(
+        @JsonProperty("enabled") Boolean enabled
+    )
+    {
+      this.enabled = enabled == null ? false : enabled;
+    }
+
+    @JsonProperty
+    final Boolean enabled;
+  }
+
+  @Override
+  public void configure(Binder binder)
+  {
+    JsonConfigProvider.bind(binder, TYPE_PREFIX, NamespaceExtractionCacheManager.class);
+    JsonConfigProvider.bind(binder, ENABLE_NAMESPACES, NamespacedExtractionModuleConfig.class);
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(JDBCExtractionNamespace.class.getCanonicalName()))
+        .to(JDBCExtractionNamespaceFunctionFactory.class)
+        .in(LazySingleton.class);
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(URIExtractionNamespace.class.getCanonicalName()))
+        .to(URIExtractionNamespaceFunctionFactory.class)
+        .in(LazySingleton.class);
+
+    binder.bind(NamespacedKeeper.class).in(ManageLifecycle.class);
+    LifecycleModule.register(binder, NamespacedKeeper.class);
```

Comment:
do we need all of
binder.bind(.).in(ManageLifecycle.class) and
LifecycleModule.register(binder, ..); and
@ManageLifecycle ? I believe either one of those should suffice.